Fix use of precedence in endpoint routing DFA (#20801)

rynowak opened the pull request:
Port of https://github.com/dotnet/aspnetcore/pull/20801

#### Description

Endpoint routing in 2.2->3.1 builds its execution graph incorrectly when a catch-all route (ex: `{**path}`) is used with higher order than a non-catchall route that overlaps it `/test/route/{id?}`.

As a result the matching behavior of routing does not match the documented behavior. The act of adding a route can cause an existing route to stop matching requests that previously worked.

The case that's impacted (using catch-all) is typical for a CMS-like "slug route" or for other frameworks like OData that integrate with routing or for a proxy.

#### Customer Impact

* Was the bug reported by a customer?

Yes. #18677 and #16579.

We've also received reports from internal partners.

* How does the bug impact the customer

Routing does not behave as documented. Requests that should match the catch-all route will not be matched. They may either be handled by lower priority routes, or "fall through" and result in a 404.

* Are there any workarounds? If so, why are they not acceptable alternatives?

Depending on the scenario it may be possible to remove or reorder routes. It might be acceptable for the user's use case, but there's no workaround to restore the documented behavior of routing. 

The other possible workaround is to disable endpoint routing and use the legacy routing system. This is a big *backwards* change for many applications. Depending on what features they are using this is not an option - Blazor Server and for instance is not supported on legacy routing.

In practice the cases that I have seen impacted are totally blocked. The user is either blocked from writing an application that functions the way they want, or they are blocked from adding features to an existing application.

#### Regression?

This is not a regression compared to the previous versions of endpoint routing,  this bug existed with endpoint routing in 2.2 and 3.0.

Compared to legacy routing this is a regression. An app that functions on 2.1 could go through our migration guide and be broken by this bug. 

#### Risk

Medium: The bug and fix are well understood and easily testable. We have a matrix of tests that runs with new and old implementations and are able to verify that the behavior is the same with 2.1 after the fix.

However, there's a risk that we could break an application by fixing the bug. Routing is a complex area and it's reasonable that some user might have tested their application and been happy with how it worked.

There's a set of possible cases where the *fixed* behavior will cause a different route to match than with the bug. Put another way the fix takes the application from a well behaved and stable (but incorrect) state to another well behaved and stable (but correct) state

-----

* Fix use of precedence in endpoint routing DFA

Fixes: #18677
Fixes: #16579

This is a change to how sorting is use when building endpoint routing's graph of
nodes that is eventually transformed into the route table. There were
bugs in how this was done that made it incompatible in some niche
scenarios both with previous implementations and how we describe the
features in the abstract.

There are a wide array of cases that might have been impacted by this
bug because routing is a pattern language. Generally the bugs will involve a
catch-all, and some something that changes ordering of templates.

Issue #18677 has the simplest repro for this, the following templates
would not behave as expected:

```
a/{*b}
{a}/{b}
```

One would expect any URL Path starting with `/a` to match the first
route, but that's not what happens.

---

The change supports an opt-out via the following AppContext switch:

```
Microsoft.AspNetCore.Routing.Use30CatchAllBehavior
```

Set to true to enable the buggy behavior for compatibility.

---

The root cause of this bug was an issue in how the algorithm used to be
build the DFA was designed. Specifically that it uses a BFS to build the
graph, and it uses an up-front one-time sort of endpoints in order to
drive that BFS.

The building of the graph has the expectation that at each level, we
will process **all** literal segments (`/a`) and then **all** parameter
segments (`/{a}`) and then **all** catch-all segments (`/{*a}`). Routing
defines a concept called *precedence* that defines the *conceptual*
order in while segments types are ordered.

So there are two problems:

- We sort based on criteria other than precedence (#16579)
- We can't rely on a one-time sort, it needs to be done at each level
(#18677)

---

The fix is to repeat the sort operation at each level and use precedence
as the only key for sorting (as dictated by the graph building algo).

We do a sort of the matches of each node *after* building the
precedence-based part of the DFA, based on the full sorting criteria, to
maintain compatibility.

rynowak committed:
Fix use of precedence in endpoint routing DFA (#20801)

* Fix use of precedence in endpoint routing DFA

Fixes: #18677
Fixes: #16579

This is a change to how sorting is use when building endpoint routing's graph of
nodes that is eventually transformed into the route table. There were
bugs in how this was done that made it incompatible in some niche
scenarios both with previous implementations and how we describe the
features in the abstract.

There are a wide array of cases that might have been impacted by this
bug because routing is a pattern language. Generally the bugs will involve a
catch-all, and some something that changes ordering of templates.

Issue #18677 has the simplest repro for this, the following templates
would not behave as expected:

```
a/{*b}
{a}/{b}
```

One would expect any URL Path starting with `/a` to match the first
route, but that's not what happens.

---

The change supports an opt-in via the following AppContext switch:

```
Microsoft.AspNetCore.Routing.UseCorrectCatchAllBehavior
```

Set to true to enable the correct behavior.

---

The root cause of this bug was an issue in how the algorithm used to be
build the DFA was designed. Specifically that it uses a BFS to build the
graph, and it uses an up-front one-time sort of endpoints in order to
drive that BFS.

The building of the graph has the expectation that at each level, we
will process **all** literal segments (`/a`) and then **all** parameter
segments (`/{a}`) and then **all** catch-all segments (`/{*a}`). Routing
defines a concept called *precedence* that defines the *conceptual*
order in while segments types are ordered.

So there are two problems:

- We sort based on criteria other than precedence (#16579)
- We can't rely on a one-time sort, it needs to be done at each level
(#18677)

---

The fix is to repeat the sort operation at each level and use precedence
as the only key for sorting (as dictated by the graph building algo).

We do a sort of the matches of each node *after* building the
precedence-based part of the DFA, based on the full sorting criteria, to
maintain compatibility.

* Add test

rynowak commented:
This is a PR against release/3.1 mostly for staging purposes and so we can review the change with the compat switch in place.

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -34,12 +36,26 @@ internal class DfaMatcherBuilder : MatcherBuilder
+            {
+                _useLegacy30Behavior = useLegacy30Behavior.Value;
+            }
+            else if (AppContext.TryGetSwitch("Microsoft.AspNetCore.Routing.Use30CatchAllBehavior", out var enabled))
```
Is it possible to add a unit test that tests setting the switch?

rynowak replied:
Not really. https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/AppContext.cs#L19

It forces tests to be single-threaded since switches are global.

JamesNK replied:
Sounds like we need an IAppContextProvider

ahah... just kidding... unless?

rynowak replied:
At some point you start staring into the void. The smaller the change the more it becomes tempting to unit test microdetails. I don't love it either, I wish that it was testable. 

We get into the problem **every** time we patch something that's worthy of an AppContext switch. Adding a unit test mitigates risk, making design changes to make things testable introduces risk.

JamesNK replied:
I was 90% [memeing](https://knowyourmeme.com/memes/just-kidding-unless) 😄 

We're always going to have flags like added in service releases. What about improving testability in 5.0? Pre-plan for 5.0.x

rynowak replied:
The switch isn't present in 5.0. No one would prefer the broken behavior.

Pilchie replied:
I think what @JamesNK was implying was maybe we should do some general work in 5.0 to thread through something that would allow us to test appcontext switches.

JamesNK replied:
Yes. If we wait until we need it (servicing) then adding appcontext switch testing will always be too risky. We can plan ahead of time in 5.0 and have it ready for future 5.0.x.

I'll make an issue to discuss more.

JamesNK reviewed (commented):
+need shiproom template

rynowak commented:
> +need shiproom template

Sure, I don't really intend for this to be ready for tactics yet. We don't have a branch for it to land in right now...

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -34,12 +36,26 @@ internal class DfaMatcherBuilder : MatcherBuilder
             _parameterPolicyFactory = parameterPolicyFactory;
             _selector = selector;
 
+            if (useLegacy30Behavior.HasValue)
```
Comment that this is for unit testing. Real world will get the value from appcontext

Pilchie commented:
:eyes: Tactics can review it ahead of time, and then we can hold it until the branch opens if you want.

rynowak commented:
> 👀 Tactics can review it ahead of time, and then we can hold it until the branch opens if you want.

I suppose that's ultimately a good thing. Since we're giving out a build of this to someone, I'd really like to give out a build that has whatever solution we decide (opt-in vs opt-out).

Pilchie commented:
Tactics approved this for 3.1.5 today, but would prefer to have it be an *opt-in* switch for 3.1 for now.  We can revisit that if we continue to hear from people affected by this who aren't finding the switch, but we should default to not changing behavior unless it's from clearly broken to working in servicing.

rynowak commented:
> 
> 
> Tactics approved this for 3.1.5 today, but would prefer to have it be an _opt-in_ switch for 3.1 for now. We can revisit that if we continue to hear from people affected by this who aren't finding the switch, but we should default to not changing behavior unless it's from clearly broken to working in servicing.

Should we set this switch in the template? I would expect we're going to hear *a lot more about this*. The fixed behavior is needed by anyone using the proxy as a library for example.

Tratcher commented:
> The fixed behavior is needed by anyone using the proxy as a library for example.

? The proxy uses 5.0.

rynowak commented:
Ok that helps a lot then.

rynowak commented:
@jamesnk - can you take a second look. I changed the pattern for the appcontext switch to be similar to our discussion, and I flipped the default behavior from fixed to broken:

- Renamed switch
- Inverted polarity
- Lots of special casing is needed now that the broken behavior is default

rynowak commented on src/Http/Routing/ref/Microsoft.AspNetCore.Routing.Manual.cs:
```diff
@@ -199,6 +199,8 @@ internal static partial class FastPathTokenizer
     {
         public DfaMatcherBuilder(Microsoft.Extensions.Logging.ILoggerFactory loggerFactory, Microsoft.AspNetCore.Routing.ParameterPolicyFactory parameterPolicyFactory, Microsoft.AspNetCore.Routing.Matching.EndpointSelector selector, System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Routing.MatcherPolicy> policies) { }
+        internal EndpointComparer Comparer { get; }
+        internal bool UseCorrectCatchAllBehavior { get; set; }
```
Note, this is an internal type. In servicing our internal types are in the ref-assemblies, and so we have to do stuff like this.

rynowak commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -52,24 +62,33 @@ internal class DfaMatcherBuilder : MatcherBuilder
-            // traverse. This means that we need to order the entries first, or else we will
-            // miss possible edges in the DFA.
-            _endpoints.Sort(_comparer);
+            if (!UseCorrectCatchAllBehavior)
```
The body of this change should be the same as the previous iteration, but with the polarity of the two ifs reversed.

rynowak commented on src/Http/Routing/test/UnitTests/Matching/RouteMatcherConformanceTest.cs:
```diff
@@ -1,10 +1,46 @@
+        [InlineData("/middleware/test", 1)]
+        [InlineData("/middleware/test1/test2", 1)]
+        [InlineData("/bill/boga", 0)]
+        public virtual async Task Match_Regression_1867(string path, int endpointIndex)
```
This test now needs to be special cased across different matchers, so it has to be duplicated.

JamesNK replied:
nit: Doesn't need to be virtual

JamesNK commented on src/Http/Routing/test/UnitTests/Matching/DfaMatcherConformanceTest.cs:
```diff
@@ -23,7 +24,132 @@ public override async Task Quirks_CatchAllParameter(string template, string path
+        [InlineData("/middleware/test", 1)]
+        [InlineData("/middleware/test1/test2", 1)]
+        [InlineData("/bill/boga", 0)]
+        public virtual async Task Match_Regression_1867_CorrectBehavior(string path, int endpointIndex)
```
nit: These tests don't need to be virtual. I don't think we'd ever inherit from DfaMatcherConformanceTest

JamesNK reviewed (approved):
Will test duplication be removable in 5.0 when the default behavior changes? If so, worth adding a comment to those tests for future generations?

rynowak commented:
> Will test duplication be removable in 5.0 when the default behavior changes? If so, worth adding a comment to those tests for future generations?

The fix is already in 5.0. This change will/should be noop merged into 5.0 and all of this code will just melt away. 

We're not porting forward the appcontext switch, etc.

rynowak commented:
This fix is ready to go once we have a branch. I've kept the no-merge label until then.

anurse commented:
Marked as approved. Noticed it was still in TacticView but was approved last meeting :).

anurse commented:
Still No Merge until 3.1.5 opens, as @rynowak said.
